### PR TITLE
feat(nuxt): warn when page uses a layout without `<NuxtLayout>`

### LINKED ISSUE #19392: Warn when a page uses a layout but no `<NuxtLayout>` is created
Following nuxt/nuxt#19388

It happens to me as well, since we have the `app.vue`, people can easily forget to add the `<NuxtLayout>`.

What if we have a variable to track if the NuxtLayout is instanciated (default to `false` and we set to `true` once created) and check if the project has layouts (including layers).

### PR DESCRIPTION
**Pull request by luc122c**
### 🔗 Linked issue

#19392

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Resolves #19392 by adding a warning in development if layouts are present but the `<NuxtLayout>` component has not been used. All changes are dev-only and should be tree-shaken out.

We first add a new variable `nuxtApp.payload._isNuxtLayoutUsed` which is `false` by default. When the `<NuxtLayout>` component is declared, this is switched to `true`. A new plugin is added which runs `onNuxtReady`. If Nuxt Layout has not been used, then it counts the layouts registered and if this is greater than 0, displays a warning. `#build/layouts` provides the layouts in the src directory, as well as in any layers.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by luc122c** [comment]
Agh, I commited to my `main` branch 😞

**Comment by luc122c** [comment]
Please could the CI be re-run, I think it's an unrelated timeout.

File: packages/nuxt/src/app/components/nuxt-layout.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import type { DefineComponent, MaybeRef, VNode } from 'vue'
 import { Suspense, Transition, computed, defineComponent, h, inject, mergeProps, nextTick, onMounted, provide, ref, unref } from 'vue'
 import type { RouteLocationNormalizedLoaded } from 'vue-router'
+import { isDevelopment } from 'std-env'
```
Comment (danielroe) [inline packages/nuxt/src/app/components/nuxt-layout.ts:4]:
```suggestion
```

File: packages/nuxt/src/app/components/nuxt-layout.ts [hunk 2]
```diff
@@ -57,6 +58,10 @@ export default defineComponent({
 
     const done = nuxtApp.deferHydration()
 
+    if (isDevelopment) {
```
Comment (danielroe) [inline packages/nuxt/src/app/components/nuxt-layout.ts:61]:
```suggestion
    if (import.meta.dev) {
```

File: packages/nuxt/src/app/nuxt.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import type { H3Event } from 'h3'
 import type { AppConfig, AppConfigInput, RuntimeConfig } from 'nuxt/schema'
 import type { RenderResponse } from 'nitropack'
 import type { MergeHead, VueHeadClient } from '@unhead/vue'
+import { isDevelopment } from 'std-env'
```
Comment (danielroe) [inline packages/nuxt/src/app/nuxt.ts:13]:
```suggestion
```

File: packages/nuxt/src/app/nuxt.ts [hunk 4]
```diff
@@ -215,7 +217,8 @@ export function createNuxtApp (options: CreateOptions) {
       data: {},
       state: {},
       _errors: {},
-      ...(import.meta.client ? window.__NUXT__ ?? {} : { serverRendered: true })
+      ...(import.meta.client ? window.__NUXT__ ?? {} : { serverRendered: true }),
+      ...(isDevelopment ? { isNuxtLayoutUsed: false } : {})
```
Comment (danielroe) [inline packages/nuxt/src/app/nuxt.ts:221]:
```suggestion
      ...(import.meta.dev ? { isNuxtLayoutUsed: false } : {})
```

File: packages/nuxt/src/app/plugins/check-if-layout-used.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import layouts from '#build/layouts'
+
+export default defineNuxtPlugin({
+  name: 'nuxt:checkIfLayoutUsed',
+  enforce: 'post',
+  setup (nuxtApp) {
+    onNuxtReady(() => {
+      if (!nuxtApp.payload.isNuxtLayoutUsed && Object.keys(layouts).length > 0) {
+        // TODO: Use logger
+        console.warn('Your project has layouts but the <NuxtLayout /> component has not been used.')
```
Comment (danielroe) [inline packages/nuxt/src/app/plugins/check-if-layout-used.ts:14]:
```suggestion
        console.warn('[nuxt] Your project has layouts but the `<NuxtLayout />` component has not been used.')
```

File: packages/nuxt/src/core/nuxt.ts [hunk 6]
```diff
@@ -7,6 +7,7 @@ import type { Nuxt, NuxtHooks, NuxtOptions } from 'nuxt/schema'
 import escapeRE from 'escape-string-regexp'
 import fse from 'fs-extra'
 import { withoutLeadingSlash } from 'ufo'
+import { isDevelopment } from 'std-env'
```
Comment (danielroe) [inline packages/nuxt/src/core/nuxt.ts:10]:
```suggestion
```

File: packages/nuxt/src/app/nuxt.ts [hunk 7]
```diff
@@ -215,7 +216,8 @@ export function createNuxtApp (options: CreateOptions) {
       data: {},
       state: {},
       _errors: {},
-      ...(import.meta.client ? window.__NUXT__ ?? {} : { serverRendered: true })
+      ...(import.meta.client ? window.__NUXT__ ?? {} : { serverRendered: true }),
+      ...(import.meta.dev ? { isNuxtLayoutUsed: false } : {})
```
Comment (danielroe) [inline packages/nuxt/src/app/nuxt.ts:220]:
```suggestion
      ...(import.meta.client ? window.__NUXT__ ?? {} : { serverRendered: true })
```

File: packages/nuxt/src/app/components/nuxt-layout.ts [hunk 8]
```diff
@@ -57,6 +57,10 @@ export default defineComponent({
 
     const done = nuxtApp.deferHydration()
 
+    if (import.meta.dev) {
+      nuxtApp.payload.isNuxtLayoutUsed = true
+    }
```
Comment (danielroe) [inline packages/nuxt/src/app/components/nuxt-layout.ts:62]:
```suggestion
    if (import.meta.dev) {
      nuxtApp.payload._isNuxtLayoutUsed = true
    }
```

File: packages/nuxt/src/app/nuxt.ts [hunk 9]
```diff
@@ -83,6 +83,7 @@ export interface NuxtPayload {
   } | null
   _errors: Record<string, NuxtError | null>
   [key: string]: unknown
+  // TODO: if isDevelopment, add isNuxtLayoutUsed
```
Comment (danielroe) [inline packages/nuxt/src/app/nuxt.ts:86]:
```suggestion
```

File: packages/nuxt/src/app/plugins/check-if-layout-used.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+// @ts-expect-error virtual file
+import layouts from '#build/layouts'
+
+export default defineNuxtPlugin({
+  name: 'nuxt:checkIfLayoutUsed',
+  enforce: 'post',
+  setup (nuxtApp) {
+    onNuxtReady(() => {
+      if (!nuxtApp.payload.isNuxtLayoutUsed && Object.keys(layouts).length > 0) {
```
Comment (danielroe) [inline packages/nuxt/src/app/plugins/check-if-layout-used.ts:12]:
```suggestion
      if (!nuxtApp.payload._isNuxtLayoutUsed && Object.keys(layouts).length > 0) {
```

File: packages/nuxt/src/app/plugins/check-if-layout-used.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+// @ts-expect-error virtual file
+import layouts from '#build/layouts'
+
+export default defineNuxtPlugin({
+  name: 'nuxt:checkIfLayoutUsed',
+  enforce: 'post',
+  setup (nuxtApp) {
+    onNuxtReady(() => {
+      if (!nuxtApp.payload._isNuxtLayoutUsed && Object.keys(layouts).length > 0) {
+        // TODO: Use logger
```
Comment (danielroe) [inline packages/nuxt/src/app/plugins/check-if-layout-used.ts:13]:
```suggestion
```

**Review by danielroe — APPROVED** [review]
This is a nice improvement 👌

**Review by TheAlexLichter — APPROVED** [review]

**Comment by nathanchase** [comment]
Question: what about if you only ever declare your layouts for your page components in `definePageMeta` - is this no long recommended? It doesn't seem "in error" if your layouts are all declared this way, is it?
```
definePageMeta({
  layout: 'user-layout',
});
```

**Comment by danielroe** [comment]
@nathanchase That should be totally fine and is recommended. What is the issue you are experiencing?

**Comment by nathanchase** [comment]
> @nathanchase That should be totally fine and is recommended. What is the issue you are experiencing?

With the nightly build, I was noticing the warning that I had layout files but didn't use `<NuxtLayout>`.

**Comment by danielroe** [comment]
Were you visiting a page with a layout defined in metadata, but not using `<NuxtLayout>`?

**Comment by nathanchase** [comment]
Correct.

**Comment by danielroe** [comment]
Well, then that's exactly what this PR is meant to warn against. You have defined a layout for a page but aren't displaying that layout. Or am I missing something?

**Comment by nathanchase** [comment]
No, I have the layout defined in the meta tag, but is that not enough to indicate that the page should be automatically wrapped by that layout? Are you meant to additionally add a `<NuxtLayout layout="user-layout"></NuxtLayout>` around your page component, as well as add it to `definePageMeta`?

**Comment by danielroe** [comment]
No, just `<NuxtLayout><NuxtPage /></NuxtLayout>` is enough. NuxtLayout defaults to the layout specified by the current route.

**Comment by nathanchase** [comment]
What about nested layouts? 

Example:
I have a `default.vue` layout in `/layouts`. 

Then I have another layout (`/layouts/user-layout.vue`) that is wrapped like so:
```
<NuxtLayout name="default">
...
</NuxtLayout>
```
Then in a page component, I reference the `user-layout` with just `definePageMeta`:
```
definePageMeta({
  layout: 'user-layout',
});
```
This has worked without issue until now.

**Comment by danielroe** [comment]
Perhaps you had better create a new issue with a reproduction. This PR did not change the actual behaviour at all, only added some console logging to avoid users _expecting_ to see a layout but not seeing one because they forgot to add `<NuxtLayout>` to their app. But that doesn't seem to be your situation.

**Comment by nathanchase** [comment]
Perhaps it is the absence of an app.vue altogether?

**Comment by danielroe** [comment]
I still don't understand the issue you're having.

**Comment by nathanchase** [comment]
The problem is that I've been using layouts for months and it all works as expected, but this PR seems to think I was doing it incorrectly and sends a warning that I'm not using `NuxtLayout`, even though I am using `<NuxtLayout name="default">` in several layout files.

**Comment by danielroe** [comment]
So you are using it.

That sounds like a bug in how we are detecting usage, then. Could you create a reproduction?

**Pull request merged**